pkg/controller/registry/resolver/satresolver.go

```diff
@@ -0,0 +1,311 @@
+package resolver
+
+import (
+	"fmt"
+
+	"github.com/blang/semver"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/controller/registry/resolver/sat"
+)
+
+type BooleanSatResolver interface {
+	SolveOperators(csvs []*v1alpha1.ClusterServiceVersion, subs []*v1alpha1.Subscription, add map[OperatorSourceInfo]struct{}) (OperatorSet, error)
+}
+
+type SatResolver struct {
+	cache OperatorCacheProvider
+}
+
+func NewDefaultSatResolver(rcp RegistryClientProvider) *SatResolver {
+	return &SatResolver{
+		cache: NewOperatorCache(rcp),
+	}
+}
+
+type installableFilter struct {
+	channel     string
+	catalog     CatalogKey
+	currentCSV  string
+	startingCSV string
+}
+
+func (s *SatResolver) SolveOperators(namespaces []string, csvs []*v1alpha1.ClusterServiceVersion, subs []*v1alpha1.Subscription, add map[OperatorSourceInfo]struct{}) (OperatorSet, error) {
+	var errs []error
+
+	installables := make([]sat.Installable, 0)
+
+	namespacedCache := s.cache.Namespaced(namespaces...)
+
+	for _, sub := range subs {
+		pkg := sub.Spec.Package
+		filter := installableFilter{
+			channel: sub.Spec.Channel,
+			catalog: CatalogKey{
+				Name:      sub.Spec.CatalogSource,
+				Namespace: sub.Spec.CatalogSourceNamespace,
+			},
+			startingCSV: sub.Spec.StartingCSV,
+			currentCSV:  sub.Status.CurrentCSV,
```

Comment:
Note to not forget - the thing we probably want here is `installedCSV`, not `currentCSV`